### PR TITLE
First pass at vector EOS interface

### PR DESCRIPTION
**Pull request by jhp-lanl**
## PR Summary

First attempt at creating a vector interface for the EOS classes. This is very much a WIP and open for discussion.

@dholladay00 and @Yurlungur please let me know what you think

### To do:
- [x] Create default vector interface
- [x] Create unit tests to cover new interface
- [x] Run tests on Darwin

## PR Checklist

- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by using the `make format` command after configuring with `cmake`.
- [x] Document any new features, update documentation for changes made.

### CODE REVIEW

File: singularity-eos/eos/eos.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos.hpp:134]: No reason to name these with the `Vector` postfix. Just let multiple dispatch handle it. Because the arguments are different, the functions will be different.

File: singularity-eos/eos/eos.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos.hpp:132]: In principle, `rhos` and `sies` could be different `Indexer` types.

File: singularity-eos/eos/eos.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos.hpp:57]: I suppose in principle `xs` and `ys` could be different indexer types.

File: singularity-eos/eos/eos.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos.hpp:335]: perhaps worth commenting that unlike the scalar versions of the functions, the lambda indexer is not optional here.

We could make it optional by adding an overload with the lambdas missing and a dummy lambda indexer passed in perhaps.

File: singularity-eos/eos/eos_variant.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos_variant.hpp:412]: Hmm... I'm not sure how I feel about re-implementing this whole thing here. Maybe it's necessary, I dunno. This whole `PTofRE` machinery might be better being refactored to reduced code duplication, and to enable individual EOSs to implement their own version if desired. (Similar to how you did the `Vector` loop abstractions.)

File: singularity-eos/eos/eos.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos.hpp:64]: This should be a `portableFor`. Look in ports of call for the syntax.

File: singularity-eos/eos/eos.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos.hpp:77]: `portableFor`.

File: singularity-eos/eos/eos_variant.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos_variant.hpp:425]: `portableFor`

File: singularity-eos/eos/eos_variant.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos_variant.hpp:434]: `portableFor`

File: singularity-eos/eos/eos_variant.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos_variant.hpp:444]: `portableFor`

File: singularity-eos/eos/eos_variant.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos_variant.hpp:449]: `portableFor`

**Review by Yurlungur — CHANGES_REQUESTED** [review]
This is a great start! Thanks for your work on this, @jhp-lanl ! A left a few organizational comments below. There's one broad thing (also discussed below) that I'd like to highlight. Mainly: the vector function calls should be called from host, but they should run on device. This means that using for loops isn't the way to go. We need to use parallel dispatch. In ports of call, we have an abstraction for this. It looks something like this:
```C++
portableFor("loop name", loop_start, loop_end, PORTABLE_LAMBDA(int i) { // loop body });
```
which wraps either an appopriate `Kokkos` call or a raw for loop depending on how the code was built. Loop bounds follow C convention, start is inclusive, end is exclusive.

Take a look at `ports-of-call/portability.hpp` for more details.

**Comment by jhp-lanl** [comment]
> There's one broad thing (also discussed below) that I'd like to highlight. Mainly: the vector function calls should be called from host, but they should run on device.

Ah okay I misunderstood you and @dholladay00 then. I thought you wanted this to be a CPU-only thing top to bottom, not that the vector functions should be host and the loops on the device. I thought about that while I was writing this but I just kept them as regular for loops for the sake of simplicity. Easy enough to change though!

I'll start addressing the individual comments here shortly.

File: singularity-eos/eos/eos_variant.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos_variant.hpp:412]: I re-implemented it simply because I was adapting what was already there. I tried to mirror the existing implementation as closely as I could just with arrays.

I think my main problem here is that different EOS may want to calculate derivatives differently and the 1e-06 factor for the perturbation is just an arbitrary number that may or may not be sufficient for all situations. I'd propose keeping the functionality similar between the scalar and vector versions of these functions, but I'm otherwise open to suggestions for to change this.

File: singularity-eos/eos/eos.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos.hpp:82]: Sounds good

File: singularity-eos/eos/eos.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos.hpp:335]: Good point. I'd prefer to mirror the functionality of the scalar functions as much as possible so I'll overload the function

**Comment by Yurlungur** [comment]
> 

Yeah it should be pretty straightforward to change.

File: singularity-eos/eos/eos_variant.hpp
Comment (Yurlungur) [reply singularity-eos/eos/eos_variant.hpp:412]: My suggestion is we should see if we can make a generic implementation that we call from each EOS, then this can be overloaded as needed.

But we can deal with that in a later PR.

File: singularity-eos/eos/eos.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos.hpp:132]: What circumstances do you think would lead to input indexers having different types? Assuming a collection of states, I'd assume that the input variables would necessarily use the same indexing scheme to travel over the collection of states.

Comment (Yurlungur) [reply singularity-eos/eos/eos.hpp:132]: Yeah you're probably right, so we can leave it for now. I was thinking of some scenario where rho and sie were sparsely populated in different ways somehow.

Comment (jhp-lanl) [reply singularity-eos/eos/eos.hpp:102]: Let's revisit if we find a circumstance where this occurs though. It will be fairly straightforward to change

**Comment by jhp-lanl** [comment]
To do:

- [ ]  Overload member functions to allow for optional specification of lambda
- [ ]  Add vector lookups to EOSPAC
- [ ]  Add unit tests of vector basic vector functionality

File: singularity-eos/eos/vector_utils.hpp
Comment (dholladay00) [inline singularity-eos/eos/vector_utils.hpp:49]: @jhp-lanl perhaps adding a `const char*` for a function name may be helpful here. Another option I've been toying with is to grab additional info from the template parameters, but getting the name of the specific function passed in so far has been challenging. I propose:
- passing in the string (lots of work added, but provides control and lots of information)
- using template param(s) (little work, but less specific, not sure if typeid overhead is perceived on every function call)

Comment (Yurlungur) [reply singularity-eos/eos/vector_utils.hpp:49]: I like the string solution better. Shouldn't be too bad.

Comment (dholladay00) [reply singularity-eos/eos/vector_utils.hpp:49]: I wrote an example of how a macro could be used that may be less objectionable to @jhp-lanl but still provide the specificity we need:
```c++
template<typename T, typename U>
void print_name(const T& a, U&& b, const char* s) {
    printf("%s\n", s);
}

#define PRINT_NAME(a, b) \
print_name(a, b, #b)

int main() {
    PRINT_NAME(gruneisen{}, &gruneisen::funcone); // string is "&gruneisen::funcone"
    return 0;
}
```

Comment (Yurlungur) [reply singularity-eos/eos/vector_utils.hpp:49]: Nice.

Comment (jhp-lanl) [reply singularity-eos/eos/vector_utils.hpp:49]: Sorry I'm a little confused. What did you want me to do with the function name?

Comment (jhp-lanl) [reply singularity-eos/eos/vector_utils.hpp]: Resolving because I changed everything lol

**Comment by jhp-lanl** [comment]
@dholladay00 and @Yurlungur I've redone everything with the CRTP design instead. I like this a lot better but as they say, the proof is in the pudding (unit tests). But in the meantime, let me know if you'd like anything changed.

File: singularity-eos/eos/eos.hpp
Comment (dholladay00) [inline singularity-eos/eos/eos.hpp:93]: could use a macro, I know preprocessor, but it could really decrease SLOC and prevent spelling errors:
```
#define SG_ADD_USINGS(MOD) \
using EosBase<MOD>::TemperatureFromDensityInternalEnergy;\
using EosBase<MOD>::InternalEnergyFromDensityTemperature;\
using EosBase<MOD>::PressureFromDensityTemperature;\
using EosBase<MOD>::PressureFromDensityInternalEnergy;\
using EosBase<MOD>::SpecificHeatFromDensityTemperature;\
using EosBase<MOD>::SpecificHeatFromDensityInternalEnergy;\
using EosBase<MOD>::BulkModulusFromDensityTemperature;\
using EosBase<MOD>::BulkModulusFromDensityInternalEnergy;\
using EosBase<MOD>::GruneisenParamFromDensityTemperature;\
using EosBase<MOD>::GruneisenParamFromDensityInternalEnergy;\
using EosBase<MOD>::FillEos;
```

Comment (jhp-lanl) [reply singularity-eos/eos/eos.hpp:93]: That's a good idea.

I'm not completely _against_ using preprocessor macros ;-) . I just don't like when they sacrifice code readability for minor gains. In this case I think the code maintenance boon is very significant and the code readability isn't sacrificed.

File: singularity-eos/eos/eos.hpp
Comment (dholladay00) [inline singularity-eos/eos/eos.hpp:1003]: Should eospac be inheriting from the base class since it's not going to be using the base class helper functions?

Comment (jhp-lanl) [reply singularity-eos/eos/eos.hpp:1003]: It shouldn't and I plan to change that. I added that as a quick workaround to get things to compile at this stage while I work on getting EOSPAC itself working in a separate way

Comment (jhp-lanl) [reply singularity-eos/eos/eos.hpp:1003]: So in short, I plan to keep the inheritance but only use the `using` statements when needed

Comment (jhp-lanl) [reply singularity-eos/eos/eos.hpp:1003]: But that reminds me that I need to actually change the scalar version of the member function in the `Variant` class... whoops 😅

**Comment by jhp-lanl** [comment]
For documentation purposes:

I was having troubles with the `ctest` framework where `ctest -V` would cause the vector test to timeout. Without the `-V` flag though, the tests work as intended.

I suspect there might be some undefined behavior occurring with the `INFO` or `REQUIRE` macros (or an interaction between the two) that is fixed when the `REQUIRE` statements are changed to `CHECK` statements.

**Comment by jhp-lanl** [comment]
> 



> For documentation purposes:
> 
> I was having troubles with the `ctest` framework where `ctest -V` would cause the vector test to timeout. Without the `-V` flag though, the tests work as intended.
> 
> I suspect there might be some undefined behavior occurring with the `INFO` or `REQUIRE` macros (or an interaction between the two) that is fixed when the `REQUIRE` statements are changed to `CHECK` statements.

It turns out that the problem was that I had an uninitialized loop counter. Catch seems to work fine so I've removed the comments that made it seem like the issue was there.

**Comment by jhp-lanl** [comment]
Okay with fixes from @dholladay00 and @jonahm-LANL all of the tests are now running and passing on Darwin Power9.

Let's do a final review tomorrow and then get this merged.

FYI - I'm going to focus on EOSPAC in a separate effort.

**Comment by Yurlungur** [comment]
Great work, @jhp-lanl ! Yes, I'll do a careful review of this tomorrow.

File: singularity-eos/eos/eos_variant.hpp
Comment (dholladay00) [inline singularity-eos/eos/eos_variant.hpp:226]: This change should probably be reverted to be consistent with the other lambdas.

Comment (jhp-lanl) [reply singularity-eos/eos/eos_variant.hpp]: Fixed

File: singularity-eos/eos/eos_base.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos_base.hpp:48]: Probably worth defining CRTP in this comment and pointing to a wikipedia article or something.

File: test/CMakeLists.txt
Comment (Yurlungur) [inline test/CMakeLists.txt:34]: What's this timeout for?

File: singularity-eos/eos/eos.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos.hpp:163]: might be worth adding a comment that this contains any generic functions provided by the base class.

**Comment by dholladay00** [comment]
pending CI and formatting, I think this is good to go!

**Review by Yurlungur — CHANGES_REQUESTED** [review]
This is a tremendous effort, @jhp-lanl ! Great work! I like the the base class solution a lot and I wonder if we can leverage it for other shared capabilities. We should think about that going forward to reduce code complexity.

Some mostly minor nitpicks below. Can you make sure to run clang-formatting before merge as well?

Also it might be good to add some comments about the design into the docs. We should maybe add a docs page for developers, explaining some of our design decisions. The compile time inheritence thing seems like the perfect place to start there. What do you think?

File: singularity-eos/eos/eos_base.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos_base.hpp:2]: Since this needs to be applied to every file in the repo, I'd like to just do this in a separate MR

Comment (Yurlungur) [reply singularity-eos/eos/eos_base.hpp:2]: I'll make an issue.

File: test/CMakeLists.txt
Comment (jhp-lanl) [reply test/CMakeLists.txt:34]: When I was dealing with undefined behavior, the loops would execute for 1500 seconds. While I could CTRL+C the tests, it didn't let the tests end to give me a clue about what was going on. Adding this timeout just makes it so each test has 60 seconds to complete instead of the default 1500.

At worst, this change does nothing at the moment and may need to be changed if somebody introduces a really long test. At best though, it will help somebody debug faster in the future if a test is misbehaving.

File: singularity-eos/eos/eos.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos.hpp:1133]: It's part of the EOSPAC vector work. I'll add a TODO

File: test/CMakeLists.txt
Comment (Yurlungur) [reply test/CMakeLists.txt:34]: ok seems fine for now.

File: singularity-eos/eos/eos_base.hpp
Comment (Yurlungur) [reply singularity-eos/eos/eos_base.hpp]: oops `member_func_name` needs to be marked static.

File: singularity-eos/eos/eos_base.hpp
Comment (dholladay00) [inline singularity-eos/eos/eos_base.hpp:313]: not necessary but could add an `#undef SG_MEMBER_FUNC_NAME` for completeness.

File: singularity-eos/eos/eos.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos.hpp]: Done

File: singularity-eos/eos/eos.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos.hpp]: Done

File: test/eos_unit_tests.cpp
Comment (jhp-lanl) [reply test/eos_unit_tests.cpp]: Done

File: singularity-eos/eos/eos_base.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos_base.hpp:53]: Added the Wiki article

File: singularity-eos/eos/eos_base.hpp
Comment (Yurlungur) [reply singularity-eos/eos/eos_base.hpp:314]: I am in favor of this.

**Review by Yurlungur — APPROVED** [review]
Approved assuming you add the `#undef` that @dholladay00 suggested. I think it would be good to add some developer documentation about this, but it doesn't have to be here---I leave that to your discretion.

Let me know when you're ready for merge @jhp-lanl .

File: singularity-eos/eos/eos_base.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos_base.hpp:316]: Sure. Added a `#undef SG_ADD_BASE_CLASS_USINGS` as well.

Comment (Yurlungur) [reply singularity-eos/eos/eos_base.hpp:316]: I don't think you can `#undef` `SG_ADD_BASE_CLASS_USINGS` because it's needed in `eos.hpp` and other header files right?

Comment (jhp-lanl) [reply singularity-eos/eos/eos_base.hpp:316]: I assumed the @dholladay00 meant adding an `#undef` right before the `#define` so I just did the same thing for `SG_ADD_BASE_CLASS_USINGS`. Did I misunderstand?

The `#undef` should just make ensure there aren't any name conflicts right?

Comment (jhp-lanl) [reply singularity-eos/eos/eos_base.hpp:316]: I'll assume that @dholladay00 wanted the `#undef` to go at the end of the file then and I'll remove both of these since they should probably be wrapped by `#ifdef`.

Comment (jhp-lanl) [reply singularity-eos/eos/eos_base.hpp:314]: Done

**Pull request merged**
